optuna/integration/_lightgbm_tuner/sklearn.py

```diff
@@ -1,49 +1,887 @@
+        min_child_weight: float = 1e-03,
+        min_child_samples: int = 20,
+        subsample: float = 1.0,
+        subsample_freq: int = 0,
+        colsample_bytree: float = 1.0,
+        reg_alpha: float = 0.0,
+        reg_lambda: float = 0.0,
+        random_state: Optional[RandomStateType] = None,
+        n_jobs: int = -1,
+        silent: bool = True,
+        importance_type: str = "split",
+        study: Optional[study_module.Study] = None,
+        timeout: Optional[int] = None,
+        train_dir: Union[pathlib.Path, str] = "optgbm_info",
+        **kwargs: Any
+    ) -> None:
+        super().__init__(
+            boosting_type=boosting_type,
+            class_weight=class_weight,
+            colsample_bytree=colsample_bytree,
+            importance_type=importance_type,
+            learning_rate=learning_rate,
+            max_depth=max_depth,
+            min_child_samples=min_child_samples,
+            min_child_weight=min_child_weight,
+            min_split_gain=min_split_gain,
+            num_leaves=num_leaves,
+            n_estimators=n_estimators,
+            n_jobs=n_jobs,
+            objective=objective,
+            random_state=random_state,
+            reg_alpha=reg_alpha,
+            reg_lambda=reg_lambda,
+            silent=silent,
+            subsample=subsample,
+            subsample_for_bin=subsample_for_bin,
+            subsample_freq=subsample_freq,
+            **kwargs,
+        )
+
+        self.study = study
+        self.timeout = timeout
+        self.train_dir = train_dir
+
+    def _check_is_fitted(self) -> None:
+        getattr(self, "n_features_")
+
+    def _get_objective(self) -> str:
+        if isinstance(self.objective, str):
+            return self.objective
+
+        if self._n_classes is None:
+            return "regression"
+        elif self._n_classes > 2:
+            return "multiclass"
+        else:
+            return "binary"
+
+    def fit(
+        self,
+        X: TwoDimArrayLikeType,
+        y: OneDimArrayLikeType,
+        sample_weight: Optional[OneDimArrayLikeType] = None,
+        init_score: Optional[OneDimArrayLikeType] = None,
+        group: Optional[OneDimArrayLikeType] = None,
+        eval_set: Optional[List[Tuple[TwoDimArrayLikeType, OneDimArrayLikeType]]] = None,
+        eval_names: Optional[List[str]] = None,
+        eval_sample_weight: Optional[List[OneDimArrayLikeType]] = None,
+        eval_class_weight: Optional[List[Union[Dict[str, float], str]]] = None,
+        eval_init_score: Optional[List[OneDimArrayLikeType]] = None,
+        eval_group: Optional[List[OneDimArrayLikeType]] = None,
+        eval_metric: Optional[Union[Callable, List[str], str]] = None,
+        early_stopping_rounds: Optional[int] = 10,
+        verbose: bool = True,
+        feature_name: Union[List[str], str] = "auto",
+        categorical_feature: Union[List[int], List[str], str] = "auto",
+        callbacks: Optional[List[Callable]] = None,
```

Comment:
Could you tell us the reason why the `callbacks` argument is passed here while `timeout` is set at `__init__`? I think we can set all of them at `fit`.

Reply:
According to the scikit-learn documentation, it is recommended not to add too many parameters to `fit_params`.

https://scikit-learn.org/stable/developers/develop.html#fitting
> Depending on the nature of the algorithm, fit can sometimes also accept additional keywords arguments. However, any parameter that can have a value assigned prior to having access to the data should be an \_\_init\_\_ keyword argument. **fit parameters should be restricted to directly data dependent variables**.

Reply:
Thank you for your comment. Then, can we move the `callbacks` to `__init__`? I don't think it is data-dependent, and I feel a bit inconsistent.

Reply:
https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMModel.html#lightgbm.LGBMModel.fit
`callbacks` is an argument that is passed to fit on the original `LGBMModel`.
If we move `callbacks` to `__init__`, it will not be compatible.

Reply:
Ah, I misunderstood. I thought this `callbacks` is Optuna's callback function, but it is actually the LightGBM's callback.
Thank you for your explanation.